stable/prometheus-operator/hack/sync_prometheus_rules.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
+"""Fetch alerting rules from provided urls into this chart."""
+import textwrap
+
+import requests
+import yaml
+from yaml.representer import SafeRepresenter
+
+
+# https://stackoverflow.com/a/20863889/961092
+class LiteralStr(str):
+    pass
+
+
+def change_style(style, representer):
+    def new_representer(dumper, data):
+        scalar = representer(dumper, data)
+        scalar.style = style
+        return scalar
+
+    return new_representer
+
+
+# Source files list
+charts = [
+    {
+        'source': 'https://raw.githubusercontent.com/coreos/prometheus-operator/master/contrib/kube-prometheus/manifests/prometheus-rules.yaml',
+        'destination': '../templates/alertmanager/rules'
+    },
+    # don't uncomment until https://github.com/etcd-io/etcd/pull/10244 is merged
+    # {
+    #     'source': 'https://raw.githubusercontent.com/etcd-io/etcd/master/Documentation/op-guide/etcd3_alert.rules.yml',
+    #     'destination': '../templates/alertmanager/rules'
+    # },
+]
+
+# Additional conditions map
+condition_map = {
+    'kube-apiserver.rules': ' .Values.kubeApiServer.enabled',
+    'kube-scheduler.rules': ' .Values.kubeScheduler.enabled',
+    'node.rules': ' .Values.nodeExporter.enabled',
+    'kubernetes-apps': ' .Values.kubeStateMetrics.enabled',
+    'etcd': ' .Values.kubeEtcd.enabled',
+}
+
+alert_condition_map = {
+    'KubeAPIDown': '.Values.kubeApiServer.enabled',  # there are more alerts which are left enabled, because they'll never fire without metrics
+    'KubeControllerManagerDown': '.Values.kubeControllerManager.enabled',
+    'KubeSchedulerDown': '.Values.kubeScheduler.enabled',
+    'KubeStateMetricsDown': '.Values.kubeStateMetrics.enabled',  # there are more alerts which are left enabled, because they'll never fire without metrics
+    'KubeletDown': '.Values.prometheusOperator.kubeletService.enabled',  # there are more alerts which are left enabled, because they'll never fire without metrics
+    'PrometheusOperatorDown': '.Values.prometheusOperator.enabled',
+    'NodeExporterDown': '.Values.nodeExporter.enabled',
+    'CoreDNSDown': '.Values.kubeDns.enabled',
+}
```

Comment:
My issue with having these depend on what is enabled vs what is not enabled is this
1. Custom scraping configurations can still make someone want to have default alerts, even though they may have disabled the component in the chart.
2. Building these switches based on the current set of alerts and recording rules from upstream means that these will be broken and require patching any time something there changes.

While not ideal, it would be a more consistent and easy to maintain experience if there are no switches.
In my case, I am throwing away all dashboards and alerts and manually configuring them via umbrella charts / other deployment mechanisms, because the defaults simply require far too much customization to be usable.

If we want to make what is available usable in a wide set of cases, we would be adding a _lot_ of configuration options to this chart. The upstream is a jsonnet repository, which assumes that the user will take the code, modify it to their requirement, then build and deploy the resources into a cluster. Helm offers far less cusomizability than that. I think it's more important to provide an easy-to-understand path with this chart than try to get closer to the jsonnet feature set.

Reply:
In my opinion it's more important to behave as expected to most users of the chart than to require disabling these altogether and making these changes local for same users. When I disabling some scrape, I don't expect to have alert about this scrape absence, and it's strainghforward what if I **do** want alert, I need to alter default configuration.
I acknowledge the problem with maintaining this import up-to-date, however with current history of change of kubernetes-mixin to me it seems insignificant: I don't know if alert names will ever be changed, for alerts which we are enclosing in `if` condition, I couldn't find it ever happened before.